ident/map.go

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ident
+
+import (
+	"github.com/m3db/m3x/checked"
+
+	"github.com/cheekybits/genny/generic"
+)
+
+const (
+	offset64 = 14695981039346656037
+	prime64  = 1099511628211
+)
+
+// ValueType is the generic value type for use with the specialized hash map.
```

Comment:
How about also accepting a generic Key type, and functions to resolve to appropriate values in the ctor.  Something like:

```
type KeyType generic.Type
type HashFn(KeyType) MapHash
type CopyFn(KeyType) KeyType
type EqualsFn(KeyType, KeyType) bool

type MapOptions struct {
  Size int64 
  HashFn HashFn
  CopyFn CopyFn
  EqualsFn EqualsFn
} 
```

Could extract that map implementation into its `m3x/generic`, and any users would specialize both key/value types

Reply:
PS this is beginning to look [familiar](http://www.cplusplus.com/reference/map/map/)

```
template < class Key,                                     // map::key_type
           class T,                                       // map::mapped_type
           class Compare = less<Key>,                     // map::key_compare
           class Alloc = allocator<pair<const Key,T> >    // map::allocator_type
           > class map;
```